### PR TITLE
feat: host-side builder-sdk via IronPythonEmbedded

### LINKED ISSUE #25: feat: make Python tooling IronPython-compatible
## Problem

The build system (`book-0-builder/builder-sdk/`) uses CPython for YAML rendering, cloud-init composition, and fragment processing. This creates a dependency boundary — PowerShell orchestrates everything but shells out to Python for the build pipeline.

IronPython runs on .NET and can be called directly from PowerShell without process boundaries, enabling:
- Direct interop between the PowerShell SDK and Python build logic
- No subprocess overhead for build operations
- Shared object model between PowerShell and Python layers
- Simpler dependency management (no separate Python install required)

### Python tools to assess for IronPython compatibility

| File | Purpose | Key dependencies |
|------|---------|-----------------|
| `builder-sdk/renderer.py` | Renders cloud-init YAML from fragments | Jinja2, PyYAML |
| `builder-sdk/context.py` | Builds template context from config files | PyYAML |
| `builder-sdk/__main__.py` | CLI entry point | argparse |
| `builder-sdk/fragments.py` | Fragment discovery and ordering | PyYAML, pathlib |

### Known IronPython considerations

- **Jinja2**: May need compatibility testing — historically has worked with IronPython but edge cases exist
- **PyYAML**: C extension (`_yaml`) won't work — must use pure Python fallback
- **pathlib**: Available in IronPython 3.x
- **argparse**: Standard library, should work
- **No C extensions**: Anything using native C modules needs a pure-Python alternative

## Open Questions

- Which IronPython version to target? (IronPython 3.4+ for Python 3 syntax)
- Should this be a full migration or dual-support (CPython + IronPython)?
- Are there any builder-sdk dependencies that are fundamentally incompatible?
- Would this change the Makefile interface or just the execution path?

### PR DESCRIPTION
**Pull request by anonhostpi**
## Goal

Move as much of the `builder-sdk/` pipeline as possible from inside the Multipass VM to the Windows host, using IronPythonEmbedded + ipy.Yaml + ipy.Jinja.

Currently the build pipeline runs inside a Multipass Ubuntu VM via CPython. By running the SDK on the host via IronPython embedded in PowerShell, we can:
- Eliminate the need to bootstrap Python inside the VM for build operations
- Run cloud-init YAML rendering, fragment composition, and template processing directly on the host
- Reduce VM provisioning time (no Python install step)
- Enable tighter integration between PowerShell orchestration and the Python build logic

## Dependencies

- [IronPythonEmbedded](https://github.com/anonhostpi/IronPythonEmbedded) — single-file IronPython embedding for PowerShell
- [ipy.Jinja](https://github.com/anonhostpi/ipy.Jinja) — Jinja2 2.10.3 + MarkupSafe 1.1.1 vendored for IronPython
- [ipy.Yaml](https://github.com/anonhostpi/ipy.Yaml) — ruamel.yaml 0.16.13 vendored for IronPython

## Scope

Assess and port the builder-sdk components:
- `builder-sdk/renderer.py` — Jinja2 YAML rendering
- `builder-sdk/context.py` — template context from config
- `builder-sdk/__main__.py` — CLI entry point
- `builder-sdk/fragments.py` — fragment discovery and ordering

Closes #25

### CODE REVIEW

**Comment by anonhostpi** [comment]
# RESEARCH: HOST.SIDE.SDK - IronPython Integration for Builder SDK

## Problem Statement

The builder-sdk (Python) currently runs inside a Multipass VM via CPython. The `Builder.Stage()` flow:

1. Creates/starts a Multipass VM (`builder`)
2. Mounts the repo into the VM (`/home/ubuntu/infra-host`)
3. Installs `python3-pip`, `python3-yaml`, `python3-jinja2`, `make` etc. via apt
4. Runs `pip3 install -e .` to install the builder package
5. Invokes `make cloud-init` / `make all` which calls `python3 -m builder render ...`
6. The VM produces rendered YAML in `output/`

**Goal:** Run the template rendering pipeline on the Windows host via IronPythonEmbedded, eliminating the VM bootstrap for build operations.

## Current Architecture

### Python SDK Call Flow

```
Makefile
  python3 -m builder render cloud-init -o output/cloud-init.yaml [--layer N]
    __main__.py -> main()
      BuildContext(config_dir)          # context.py
        Loads all *.config.yaml from book-1/book-2
        Auto-unwraps single-key files
        Applies testing overrides (deep_merge)
        Loads OAuth credential fallbacks from host files
        Applies AUTOINSTALL_* env var overrides

      render_cloud_init_to_file(ctx, output, ...)  # renderer.py
        discover_fragments()
          Scans book-1/book-2 for build.yaml files
          Returns sorted by build_order

        render_scripts(ctx)
          For each fragment with scripts/ dir:
            Render *.sh.tpl via Jinja2 -> dict of script content

        For each matching fragment:
          Render fragment.yaml.tpl via Jinja2
            (template context = config + scripts dict)
          yaml.safe_load() the rendered text

        deep_merge() all fragment dicts

        yaml.dump() -> #cloud-config header + merged YAML
```

### PowerShell SDK Invocation Path

```
Invoke-IncrementalCloudInitTest.ps1
  . SDK.ps1  (loads all modules)
    $SDK.Builder.Stage()
      Setup(true)       -> Create Multipass VM
      Mount(root)       -> Mount repo into VM
      InstallDependencies()
        apt-get install python3-pip python3-yaml python3-jinja2 make ...
        pip3 install -e .
    $SDK.CloudInit.Test.Run($Layer)
      $SDK.Builder.Runner($config, "Multipass", $Layer)
        $this.Build($Layer)  -> make cloud-init LAYER=$Layer (in VM)
```

### Files in builder-sdk/

| File | Lines | Purpose | Dependencies |
|------|-------|---------|-------------|
| `__init__.py` | 1 | Package marker | - |
| `__main__.py` | 173 | CLI entry point (argparse) | context, renderer, artifacts |
| `context.py` | 444 | Config loading, env overrides, credential fallbacks | yaml, json, os, re, pathlib, composer |
| `renderer.py` | 233 | Jinja2 rendering, fragment discovery, deep merge | jinja2, yaml, pathlib, artifacts, filters, composer |
| `composer.py` | 26 | `deep_merge()` utility | - (pure Python) |
| `filters.py` | 151 | Jinja2 custom filters (base64, yaml, sha512, shell, network) | hashlib, base64, os, yaml |
| `artifacts.py` | 66 | Build manifest tracking (artifacts.yaml) | pathlib, datetime, yaml |

## Portability Assessment: IronPython 3.4 Compatibility

### Syntax Issues

| Issue | Location | Severity | Fix |
|-------|----------|----------|-----|
| `str.removesuffix()` (Python 3.9+) | `renderer.py:97` | **Blocker** | Replace with `s[:-len(suffix)] if s.endswith(suffix) else s` |
| f-strings | Multiple files | **OK** | IronPython 3.4 backported f-string support |

### Standard Library Concerns

| Module | Used In | IronPython Status | Notes |
|--------|---------|-------------------|-------|
| `pathlib` | context, renderer, artifacts | **Likely OK** | Part of Python 3.4 stdlib, but IronPython may have filesystem quirks |
| `hashlib` | filters.py | **Likely OK** | IronPython wraps .NET crypto; SHA-512 should work |
| `os.urandom` | filters.py | **Likely OK** | Maps to .NET crypto RNG |
| `json` | context.py | **OK** | Pure Python stdlib |
| `re` | context.py | **OK** | Pure Python stdlib |
| `base64` | filters.py | **OK** | Pure Python stdlib |
| `datetime` | artifacts.py | **OK** | Pure Python stdlib |
| `argparse` | __main__.py | **Not needed** | CLI entry point replaced by PowerShell wrapper |
| `yaml` (PyYAML) | Multiple | **Replaced** | Use `ruamel.yaml` via ipy.Yaml instead |
| `jinja2` | renderer.py | **Replaced** | Use Jinja2 2.10.3 via ipy.Jinja |

### Jinja2 Version Compatibility

**Current**: pyproject.toml requires `jinja2>=3.0`
**IronPython**: ipy.Jinja vendors Jinja2 **2.10.3**

Key API differences:
- `Environment`, `FileSystemLoader`, `DictLoader` -- all available in 2.10.3
- Template rendering API (`render()`, `get_template()`) -- same
- Custom filters registration -- same
- `keep_trailing_newline` parameter -- available since 2.7

**No breaking changes expected** for the features used by this SDK.

### PyYAML vs ruamel.yaml

The SDK uses `yaml.safe_load()`, `yaml.dump()`, and `yaml.add_representer()`. The ruamel.yaml API differs:

| PyYAML | ruamel.yaml 0.16 | Used In |
|--------|-------------------|---------|
| `yaml.safe_load(s)` | `YAML().load(s)` | context.py, renderer.py |
| `yaml.dump(d, f, ...)` | `YAML().dump(d, f)` | renderer.py, artifacts.py, filters.py |
| `yaml.add_representer(str, fn)` | Custom `Representer` subclass | renderer.py |

This is a significant refactoring point. The SDK needs a YAML abstraction layer.

### FileSystemLoader Concern

The current code uses `FileSystemLoader(['book-1-foundation', 'book-2-cloud'])` with relative paths. Under IronPython on the host, this will use Windows filesystem paths. The template references use POSIX-style paths (`tpl_path.as_posix()`).

**Resolution**: Use absolute paths from the repo root. `FileSystemLoader` accepts absolute paths and should work on Windows with IronPython since it uses real filesystem I/O (not in-memory).

### Credential Fallback Code

`context.py` has methods that read host-side credential files (`~/.claude/.credentials.json`, `~/.copilot/config.json`). These currently run inside the VM where the files don't exist (silently skipped). On the host, they **will** find and load these files, which is actually the desired behavior.

## Proposed Architecture

### Design: Two-Layer Approach

Rather than porting the Python SDK directly to IronPython (fighting API differences), create a **PowerShell rendering module** that uses IronPython as a Jinja2+YAML execution engine.

```
book-0-builder/
  builder-sdk/           # Existing Python SDK (kept for VM fallback)
  host-sdk/
    modules/
      Renderer.ps1       # NEW: IronPython-based rendering module
      ...
    SDK.ps1              # Modified: loads Renderer module
  ipy/                   # NEW: IronPython rendering engine
    engine.ps1           # Engine setup (IronPythonEmbedded + ipy.Jinja + ipy.Yaml)
    renderer.py          # Python rendering logic (loaded into IronPython)
    filters.py           # Python filters (loaded into IronPython)
```

### Component Design

#### 1. `book-0-builder/ipy/engine.ps1` -- Engine Bootstrap

Downloads and initializes IronPythonEmbedded + ipy.Jinja + ipy.Yaml. Returns an engine object ready to execute rendering.

```powershell
function New-RenderEngine {
    param([string]$RepoRoot)
    $builder = iwr 'https://raw.githubusercontent.com/.../IronPythonEmbedded.ps1' | iex
    $engine = $builder.Build()
    . "$PSScriptRoot/../ipy.Jinja/ipy.Jinja.ps1"
    Install-IpyJinja -Engine $engine
    . "$PSScriptRoot/../ipy.Yaml/ipy.Yaml.ps1"
    Install-IpyYaml -Engine $engine
    # Load the rendering module
    $engine.Add("/ipy/sdk/renderer.py", (Get-Content "$PSScriptRoot/renderer.py" -Raw))
    $engine.Add("/ipy/sdk/filters.py", (Get-Content "$PSScriptRoot/filters.py" -Raw))
    return $engine
}
```

#### 2. `book-0-builder/ipy/renderer.py` -- IronPython Rendering Logic

A streamlined version of the builder-sdk, adapted for:
- ruamel.yaml instead of PyYAML
- Jinja2 2.10.3 API
- No argparse (called from PowerShell, not CLI)
- No `removesuffix()` (Python 3.9+)
- Paths passed in from PowerShell (not discovered via `Path.rglob`)

Key functions exposed to PowerShell:
- `render_cloud_init(repo_root, config_data, layer, include, exclude)` returns YAML string
- `render_autoinstall(repo_root, config_data)` returns YAML string

#### 3. `book-0-builder/host-sdk/modules/Renderer.ps1` -- PowerShell Wrapper

SDK module that:
1. Initializes the IronPython engine (lazy, first use)
2. Discovers fragments (reuse existing Fragments.ps1 logic)
3. Reads config (reuse existing Config.ps1 / Build-TestConfig)
4. Passes config + template paths to IronPython for rendering
5. Writes output files

API: `$SDK.Renderer.RenderCloudInit($Layer, $OutputPath)` and `$SDK.Renderer.RenderAutoinstall($OutputPath)`

### Data Flow (Proposed)

```
PowerShell Host
  |
  +-- Build-TestConfig()          # Config.ps1 (already exists)
  |     *.config.yaml -> hashtable
  |
  +-- $SDK.Fragments.UpTo($Layer) # Fragments.ps1 (already exists)
  |     build.yaml discovery -> fragment list
  |
  +-- IronPython Engine
        Receives: config dict, fragment paths, template paths
        Jinja2 renders each fragment.yaml.tpl
        ruamel.yaml parses rendered YAML
        deep_merge combines fragments
        Returns: rendered YAML string -> PowerShell writes to file
```

### What Gets Ported vs Reused

| Component | Action | Reason |
|-----------|--------|--------|
| `composer.py` (deep_merge) | **Port to IronPython** | Pure Python, trivial |
| `filters.py` | **Port to IronPython** | Needs ruamel.yaml swap for `to_yaml` filter |
| `renderer.py` | **Rewrite for IronPython** | Drop CLI/pathlib/discovery, accept data from PS |
| `context.py` | **Skip** | Config loading already done by PowerShell Config.ps1 |
| `__main__.py` | **Skip** | CLI replaced by PowerShell module |
| `artifacts.py` | **Port or skip** | Simple; could stay in PowerShell |
| `Config.ps1` | **Reuse as-is** | Already loads all *.config.yaml |
| `Fragments.ps1` | **Reuse as-is** | Already discovers fragments |

### Eliminating the Builder VM

With host-side rendering, `Builder.Stage()` flow changes:

**Before (VM required for rendering):**
1. Create VM -> Mount repo -> Install Python deps -> `make cloud-init`

**After (host-side rendering):**
1. PowerShell loads config and discovers fragments
2. IronPython renders templates on host
3. VM only needed for: ISO building (`xorriso`), testing (running cloud-init)

The Builder VM is still needed for ISO building and as a test runner, but **not for template rendering**.

## Files to Modify

| File | Change |
|------|--------|
| `book-0-builder/ipy/engine.ps1` | **NEW** -- IronPython engine bootstrap |
| `book-0-builder/ipy/renderer.py` | **NEW** -- Streamlined rendering for IronPython |
| `book-0-builder/ipy/filters.py` | **NEW** -- Adapted filters (ruamel.yaml, no removesuffix) |
| `book-0-builder/ipy/composer.py` | **NEW** -- deep_merge (copy from existing) |
| `book-0-builder/host-sdk/modules/Renderer.ps1` | **NEW** -- PowerShell rendering module |
| `book-0-builder/host-sdk/SDK.ps1` | **MODIFY** -- Load Renderer module |
| `book-0-builder/host-sdk/modules/Builder.ps1` | **MODIFY** -- Use host rendering when available, VM fallback |
| `book-0-builder/host-sdk/modules/CloudInit.ps1` | **MODIFY** -- Add rendering methods |

## Blockers and Open Questions

### Confirmed Blockers
1. **`str.removesuffix()`** in renderer.py -- Python 3.9+ only. Easy fix.

### Potential Blockers (Need Testing)
2. **pathlib on IronPython** -- Used for template path manipulation. May need to fall back to `os.path`. Standard library since 3.4 but IronPython may have incomplete implementation.
3. **hashlib SHA-512** -- Used for password hashing in filters.py. Should work via .NET crypto, but the custom `_sha512_crypt` algorithm (glibc-compatible) needs verification.
4. **`os.urandom()`** -- Used for salt generation. Should map to .NET RNG.
5. **Jinja2 FileSystemLoader with Windows paths** -- Templates use POSIX paths internally; need to verify cross-platform path handling.

### Design Questions
6. **Submodule vs download**: Should IronPythonEmbedded/ipy.Jinja/ipy.Yaml be git submodules, or downloaded at runtime (as their READMEs show)?
7. **YAML representer**: The custom `str_representer` that forces literal block scalars for multiline strings -- need to verify ruamel.yaml equivalent approach. ruamel.yaml 0.16 has built-in support for literal block scalars via `yaml.default_style` and `ScalarString` types.
8. **Config passing**: PowerShell to IronPython data marshalling. The engine's `scope.SetVariable()` / `scope.GetVariable()` can pass .NET objects, but complex nested dicts may need JSON serialization as an intermediate format.

### Not Blockers
- **Credential fallbacks** -- Will actually work *better* on host (files are present)
- **f-strings** -- Supported in IronPython 3.4
- **Jinja2 2.10.3 vs 3.0** -- No breaking API differences for features used

File: book-0-builder/ipy/engine.ps1 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#endregion
+
+#region MARK: Load ipy.Yaml
+$ipy_yaml_loader = Invoke-WebRequest 'https://raw.githubusercontent.com/anonhostpi/ipy.Yaml/main/ipy.Yaml.ps1' -UseBasicParsing
+& ([scriptblock]::Create($ipy_yaml_loader.Content)) -Engine $ipy_engine
+#endregion
+
+#region MARK: Load ipy.Jinja
+$ipy_jinja_loader = Invoke-WebRequest 'https://raw.githubusercontent.com/anonhostpi/ipy.Jinja/main/ipy.Jinja.ps1' -UseBasicParsing
+& ([scriptblock]::Create($ipy_jinja_loader.Content)) -Engine $ipy_engine
```
Comment (anonhostpi) [inline book-0-builder/ipy/engine.ps1:18]: I'm pretty sure this is wrong

**Comment by anonhostpi** [comment]
from claude:
> **Rule 6 Research — renderer.py incremental patching**
> 
> ## Problem
> 
> The PR diff for `renderer.py` shows 232 deletions + 86 additions (no rename detection) because the file's git lineage was broken during development. Three attempts were made:
> 
> 1. **Attempt 1** (`3a92a3b`-`08d8689`): Written from scratch, then reverted
> 2. **Attempt 2** (`336d0d9`-`d357d0c`): Proper `git mv` + incremental patches, but reverted as "incorrectly patched"
> 3. **Attempt 3** (`941b79d`-`9943b48`): File *copied* (not `git mv`) into `ipy/`, then incrementally patched — current state
> 
> Attempt 2 used `git mv` correctly (commit `336d0d9` shows `0 insertions, 0 deletions` — a pure rename). But attempt 3 at commit `941b79d` re-added the file with 232 insertions, breaking rename tracking. Since `builder-sdk/renderer.py` was already deleted in attempt 2, git cannot connect the two.
> 
> ## Functional Equivalence Verification
> 
> The current 86-line `ipy/renderer.py` is functionally correct:
> 
> | Original (232 lines) | Current (86 lines) | Status |
> |---|---|---|
> | `from pathlib import Path` / `import yaml` / `from .` imports | `import os` / `from ruamel.yaml import YAML` / direct imports + `sys`, `json` | Correct |
> | `str_representer` + `yaml.add_representer` | Removed (ruamel.yaml handles natively) | Correct |
> | `render_text`, `render_scripts`, `render_script` | Removed (moved to PowerShell) | Correct |
> | `get_available_fragments`, `FragmentValidationError` | Removed (moved to PowerShell) | Correct |
> | `get_environment`, `_env` global | Removed (env created in `main()`) | Correct |
> | `render_cloud_init_to_file`, `render_autoinstall`, `render_autoinstall_to_file` | Removed (moved to PowerShell) | Correct |
> | `discover_fragments(base_dirs)` | `discover_fragments(repo_root, base_dirs)` with `os.walk` + `YAML().load()` | Correct |
> | `create_environment(template_dirs)` | `create_environment(repo_root, template_dirs)` with `abs_dirs` | Correct |
> | `render_cloud_init(ctx, include, exclude, layer, for_iso)` | `render_cloud_init(ctx, env, fragments)` — filtering moved to PowerShell | Correct |
> | N/A | `main()` — reads JSON stdin, writes YAML stdout | New entry point, correct |
> | No IronPython concerns | No f-strings, no `.removesuffix()`, no `pathlib` | Correct |
> 
> ## Fix: Squash and Redo with Rename Tracking
> 
> The incremental patch commits (attempt 3: `92e1373`-`9943b48`) have the right *content* but broken *rename lineage*. The fix requires:
> 
> ### Option A: Squash everything, redo as `git mv` + patches (recommended)
> 
> 1. Reset the renderer.py changes back to master state (restore `builder-sdk/renderer.py`, remove `ipy/renderer.py`)
> 2. Start with `git mv book-0-builder/builder-sdk/renderer.py book-0-builder/ipy/renderer.py` (pure rename commit)
> 3. Apply the same incremental patch sequence from attempt 3 — the commit messages and content are already correct
> 4. GitHub will show the diff as a rename+modify with `~37% similarity` rather than delete+create
> 
> Using `-M10%` threshold, git already detects the rename:
> ```
> similarity index 10%
> rename from book-0-builder/builder-sdk/renderer.py
> rename to book-0-builder/ipy/renderer.py
> ```
> 
> ### Option B: Interactive rebase to fix history (risky)
> 
> Replace commits `d357d0c` + `941b79d` with a single commit that does `git mv` instead of delete+copy. This preserves the incremental patch commits but requires careful rebasing across ~20 commits.
> 
> ## Recommended Approach
> 
> **Option A**. The patch sequence is well-defined (18 commits from `92e1373` to `9943b48`), the content is verified correct, and the commit messages can be reused verbatim. The only change is ensuring the initial "move" commit uses `git mv` so GitHub's PR view shows a proper rename diff.
> 
> ## Files to Modify
> 
> | File | Change |
> |------|--------|
> | `book-0-builder/builder-sdk/renderer.py` | `git mv` to `ipy/renderer.py` (not delete+create) |
> | `book-0-builder/ipy/renderer.py` | Same incremental patches as attempt 3 |

**Comment by anonhostpi** [comment]
Closing in favor of a redo with proper git mv tracking for all Python files.
